### DISCUSSION TITLE
aes(!!!GeomPoint$default_aes, !!!GeomSegment$default_aes) without warning

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#30] [Q&A]
Is there any way to combine existing Geom aes without warning when there are overlaps?

``` r
library(ggplot2)
aes(!!!GeomPoint$default_aes, !!!GeomSegment$default_aes)
#> Warning: Duplicated aesthetics after name standardisation: colour and alpha
#> Aesthetic mapping: 
#> * `shape`     -> 19
#> * `colour`    -> "black"
#> * `size`      -> 1.5
#> * `fill`      -> NA
#> * `alpha`     -> NA
#> * `stroke`    -> 0.5
#> * `colour`    -> "black"
#> * `linewidth` -> 0.5
#> * `linetype`  -> 1
#> * `alpha`     -> NA
```

``` r

aes(!!!GeomPoint$default_aes, !!!GeomSegment$default_aes)[c(-2,-5)] 
#> Warning: Duplicated aesthetics after name standardisation: colour and alpha
#> Aesthetic mapping: 
#> * `shape`     -> 19
#> * `size`      -> 1.5
#> * `fill`      -> NA
#> * `stroke`    -> 0.5
#> * `colour`    -> "black"
#> * `linewidth` -> 0.5
#> * `linetype`  -> 1
#> * `alpha`     -> NA
```

<sup>Created on 2024-07-21 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by yjunechoe** [comment]
One way is to use `modifyList()`, though it resolves overlaps asymmetrically. This preserves all default aesthetic mappings from GeomPoint and adds extra ones found in GeomSegment. Might need extra care if you imagine there being two mappings to the same aes using different default values.

```r
library(ggplot2)
aes(!!!modifyList(GeomPoint$default_aes, GeomSegment$default_aes))
#> Aesthetic mapping: 
#> * `shape`     -> 19
#> * `colour`    -> "black"
#> * `size`      -> 1.5
#> * `fill`      -> NA
#> * `alpha`     -> NA
#> * `stroke`    -> 0.5
#> * `linewidth` -> 0.5
#> * `linetype`  -> 1
```

**Reply by EvaMaeRey** [reply]
This looks great!  Asymmetric and keeping the first seems like a good behavior.

**Reply by EvaMaeRey** [reply]
I'm using it over in #26, which started by looking at the statexpress package usage, but I'd be interested in your thoughts no-dependency proposal: https://github.com/teunbrand/ggplot-extension-club/discussions/26#discussioncomment-10024804

**Comment by EvaMaeRey** [comment]
I'm back here thinking about a simpified one-off replacement for specific aes, like:

GeomPoint2 <- ggproto("GeomPoint2", GeomPoint, default_aes = aes(!!!modifyList(aes(color = "orange"), GeomPoint$default_aes))

A shortcut from 'inheriting from an existing Geom' which in itself is of course already a huge shortcut. 

GeomPolygonHollow <- ggproto("GeomPolygonHollow", GeomPolygon,
  default_aes = aes(colour = "black", fill = NA, linewidth = 0.5, linetype = 1,
    alpha = NA)
  )

But it seems a little expensive in that you might just want to modify a specific thing, but have to copy paste all the existing specifications.


So maybe worth a little explore and following qstat construction ideas, but not building a Geom from scratch. https://github.com/EvaMaeRey/statexpress

qgeom_mod <- function(geom, default_aes_new){

ggproto("GeomTemp", geom, default_aes = aes(!!!modifyList(default_aes_new, geom$default_aes))

}

I do wonder if the special replacement would be protected (unchanged) from geom theming decisions available in a ggplot2 release.  @teunbrand 

My bet would be yes (maybe that's hope speaking because it seems like a preferable outcome), especially if used like...

geom_point_mod <- function(...){

QGeomPoint <- qgeom_mod(GeomPoint, aes(color = "orange"))

stat_identity(geom = QGeomPoint, ...)

}

**Reply by yjunechoe** [reply]
After seeing some usecases for splicing default_aes here, I have a question: how does this approach differ from the more "canonical" approach of doing this at the level of the constructor function (vs. ggproto object)?

So, to compare, I assume the advantage of what you propose here is that you can have a partially modified `Geom` that you can pair up with different stats:

```r
library(ggplot2)
df <- data.frame(x = 1, y = 1:10)

GeomPoint2 <- ggproto(
  "GeomPoint2", GeomPoint,
  default_aes = aes(!!!modifyList(GeomPoint$default_aes, aes(color = "orange")))
)

ggplot(df, aes(x, y)) +
  stat_identity(geom = GeomPoint2)

ggplot(df, aes(x, y)) +
  stat_summary(geom = GeomPoint2)
```

But you could also have a partial `geom_*()` function that hard codes the color into the instance of the layer, which still allows the mix-and-match

```r
geom_point_orange <- function(...) {
  geom_point(..., color = "orange")
}

ggplot(df, aes(x, y)) +
  geom_point_orange(stat = "identity")

ggplot(df, aes(x, y)) +
  geom_point_orange(stat = "summary")
```

Do you imagine this approach/`qgeom_mod` to serve other purposes?

BTW something like a `qgeom_mod` for the layer approach would be to construct a partial function, which I think is kinda slick for different reasons :)

```r
geom_point_orange <- purrr::partial(geom_point, color = "orange")

ggplot(df, aes(x, y)) +
  geom_point_orange(stat = "identity")
```

**Reply by EvaMaeRey** [reply]
Thanks for the ideas, @yjunechoe!  The mix and match layer thinking is ✨✨✨!

I was motivated by some experiences here, long story short, I didn't want hard code the Geom aes: 

https://evamaerey.github.io/mytidytuesday/2024-09-20-corr-w-qstat-qlayer/corr-w-qstat-qlayer.html


I just wanted to change the color to black for the GeomRect, but didn't want to lock that in, otherwise I would have used geom_rect(stat = QStat, ...) to define my convenience layer.  Mostly I think not messing with the defaults isn't a great idea - one more thing to manage - but this geom_ is meant to be used with other geoms, and it was just blends in with the previous (also Rect) isn't a color doing some work to define the border.  

qstat is ggproto("StatTemp", Stat, compute_group,...)
and qlayer is like geom_point but geom argument is available.

geom_covariance <- function(...){
  
  QStat <- qstat(compute_covariance,
                 default_aes = aes(fill = after_stat(sign)))
  
  GeomRect2 <- ggproto("GeomRect2", GeomRect, 
                       default_aes = aes(color = "black", 
                              fill = "grey35",
                              linewidth = .5,
                              linetype = 1, 
                              alpha = NA))
  
  qlayer(stat = QStat, geom = GeomRect2, ...)
  
}  

and changing GeomHline to dashed

# SDs
compute_sdy <- function(data, scales){
  
  ymean <- mean(data$y)
  ysd <- sd(data$y)
  
  data.frame(yintercept = ymean + ysd*c(-2,-1,1,2))

    }


geom_y_sd <- function(...){
  
  QStat <- qstat(compute_sdy, 
              dropped_aes = c("x", "y"))
  
  # GeomHline$default_aes
  GeomHline2 <- ggproto("GeomHline2", GeomHline, 
                        default_aes = aes(colour = "black",
                                          linewidth = .5,
                                          linetype = "dotted",
                                          alpha = NA))
  
  qlayer(geom = GeomHline2, stat = QStat, ...)
  
}

Only did the process a few times, but feels like a little more work than it should be (first you have to check out GeomRect$default_aes), and reading it back it's also not obvious how GeomRect2 is different from GeomRect.  So ModifyList and !!! to the rescue?  Thoughts?

I'm guessing the original question was for use with something like a lollipop combo geom - but I don't even remember!

purrr::partial looks very useful!

**Reply by teunbrand** [reply]
The shortcoming of hardcoding the colour in the geom, is that the user cannot override it. Even if you make it an argument, the user might want to map something to that aesthetic, instead of having it static. The default aes route allows for the usual customisations.

With regards to verbosity, nothing is preventing you to write a little helper to take care of the gnarly bits.

``` r
library(ggplot2)
with_defaults <- function(new, default) {
  new <- c(new, default[setdiff(names(default), names(new))])
  default[names(new)] <- new
  default
}
with_defaults(
  aes(colour = "orange"), 
  default = aes(fill = "black", colour = "red")
)
#> Aesthetic mapping: 
#> * `fill`   -> "black"
#> * `colour` -> "orange"
```

<sup>Created on 2024-10-03 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
Ahhh got it - thanks @EvaMaeRey for the explanation, and good point @teunbrand!

I was thinking too narrowly, like "Why would you want a `GeomPointOrange` with the ability to override the color to be anything other than orange?" But I see now how you might want that flexibility if _multiple_ defaults are at play simultaneously (with `GeomRect2` being a better example to this point)

**Reply by EvaMaeRey** [reply]
I'm pretty sure I've written a 'geom_point_blue' at some point and direct path makes sense!

I think I don't really understand the with_defaults proposal.  It could be used instead of the !!! + modifyList approach.  Is that it?

**Reply by EvaMaeRey** [reply]
All's clear. 

``` r
library(tidyverse)

with_defaults <- function(new, default) {
  new <- c(new, default[setdiff(names(default), names(new))])
  default[names(new)] <- new
  default
  }

GeomPoint2 <- ggproto("GeomPoint2", 
                       GeomPoint, 
                       default_aes = with_defaults(aes(color = "blue"), GeomPoint$default_aes))


ggplot(cars) + 
  aes(speed, dist) + 
  layer(stat = StatIdentity, 
        geom = GeomPoint2,
        position = "identity")
```

![](https://i.imgur.com/sVp4a59.png)

<sup>Created on 2024-10-03 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I still don't have a sense of whether bespoke decisions like this would get modified by the future theming that you'll be able to do to geoms.   My thought is that in the above construction "blue" might be overriden.

But 'blue' might render here:

ggplot(cars) + 
  aes(speed, dist) + 
  layer(stat = StatIdentity, 
        geom = ggproto("GeomPoint2", GeomPoint, with_defaults(aes(color = "blue"), GeomPoint$default_aes)),
        position = "identity")

I'll have to check out the branch!

**Reply by teunbrand** [reply]
In the future, most `default_aes` field are soft-coded to retrieve their value from the theme. If you hardcode let's say `colour = "blue"`, it will not be overriden by the theme.

**Reply by EvaMaeRey** [reply]
That sounds great.

**Reply by EvaMaeRey** [reply]
I like the idea of both...

qgeom(`_inherit` = GeomPoint, 
              default_aes = default_aes_update(existing = GeomPoint$default_aes, mod = aes(color = "blue")) 

And the wrapped up...

qggproto_update(`_inherit` = GeomPoint, default_aes_update = aes(color = "blue")) # q for 'quick'

So also usable with Stat default updates - like the likert StatProp default switches that have recently been discussed in #36.

**Reply by larmarange** [reply]
I also like the idea of qggproto_update function to facilitate the update. 

Just for info, I tried in ggstat to make a copy of GeomText with a default aes colour equal to after_stat(hex_bw(.data$fill)). It never worked with modifyList approach. I got an error that .data was used in an inappropriate context. However, it works perfectly if I manually pass this value with aes() when constructing my plot. I have to admit that I didn't identify where the error was coming from.

**Reply by EvaMaeRey** [reply]
I sketched `proto_update` and `qproto_update` (on-the-fly definitions) in statexpress, https://github.com/EvaMaeRey/statexpress .  I didn't feel like default_aes_update added much value as a stand alone, but include it below for reference.  

Syntax highlight:

``` r
ggplot(cars) + 
  aes(speed, dist) + 
    layer(geom = qproto_update(GeomPoint, aes(color = "blue")),
          stat = StatIdentity,
          position = "identity")
```

![](https://i.imgur.com/rqMEXl6.png)


Details

``` r
library(tidyverse)


proto_update <- function(`_class`, `_inherit`, default_aes_update = NULL, ...){
  
  if(!is.null(default_aes_update)){
  
    default_aes <- aes(!!!modifyList(`_inherit`$default_aes, default_aes_update))
    
    }
  
  ggproto(`_class`, `_inherit` = `_inherit`, default_aes = default_aes, ...)
  
}

GeomPointCornflowerblue <- proto_update("GeomPointCornflowerblue", 
                                    GeomPoint,
                                    default_aes_update = aes(color = "cornflowerblue"))

ggplot(cars) + 
  aes(speed, dist) + 
    layer(geom = GeomPointCornflowerblue,
          stat = StatIdentity,
          position = "identity")
```

![](https://i.imgur.com/j5Q8woy.png)




``` r

qproto_update <- function(`_inherit`, default_aes_update = NULL, ...){
  
  proto_update("protoTemp", 
               `_inherit`, 
               default_aes_update = default_aes_update,
               ...)
  
}

ggplot(cars) + 
  aes(speed, dist) + 
    layer(geom = qproto_update(GeomPoint, aes(color = "blue")),
          stat = StatIdentity,
          position = "identity")
```

![](https://i.imgur.com/rqMEXl6.png)

``` r


default_aes_update <- function(`_inherit`, update){
  
  aes(!!!modifyList(`_inherit`$default_aes, update))
  
}


GeomPointPink <- ggproto("GeomPointPink", GeomPoint, 
                         default_aes = 
                           default_aes_update(GeomPoint, aes(color = "pink")))

ggplot(cars) + 
  aes(speed, dist) + 
    layer(geom = GeomPointPink,
          stat = StatIdentity,
          position = "identity")
```

![](https://i.imgur.com/LhAvqxI.png)

<sup>Created on 2024-10-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I think @yjunechoe might be able to nicely articulate why `after_stat(hex_bw(.data$fill))` doesn't work when constructing a stat but apparently works in an active plot.  Resident after_stat() expert. And might have a work around.  That said, I think you are in a pretty good place without automatically adjusting the text colors.

**Reply by larmarange** [reply]
By the way, my mistake, this is an issue with after_scale and not after_stat, i.e `after_scale(hex_bw(.data$fill))`

**Reply by EvaMaeRey** [reply]
after_scale makes sense. I'm not familiar with Stats that define default_aes w/ after_scale. I wonder if it possible given plot build pipeline.  

A not baked-in approach, could be something like aes_color_after_fill(). And maybe that would work like this:

ggplot(survey) + 
  aes(y = question, fill = response) + 
  geom_likert() + 
  geom_likert_text() + 
  aes_color_after_fill()

or if you prefer, 

ggplot(survey) + 
  aes(y = question, fill = response, !!!aes_color_after_fill()) + 
  geom_likert() + 
  geom_likert_text()

Using ggplot tricks base contrasting code (but you could probably use hex_bw similarly), https://github.com/teunbrand/ggplot_tricks

contrast <- function(colour) {
  out   <- rep("black", length(colour))
  light <- farver::get_channel(colour, "l", space = "hcl")
  out[light < 50] <- "white"
  out
}

aes_color_after_fill <- function(fun = contrast){
aes(colour = after_scale(fun(fill)))
}

**Reply by larmarange** [reply]
Thanks for the trick. It looks great. I have tried to implement `autocontrast` and it works well. cf. https://github.com/larmarange/ggstats/pull/76

Except in one case. If I try to create:

```r
GeomTextAutoContrast <- ggproto(
  "GeomTextAutoContrast",
  ggplot2::GeomText,
  default_aes = ggplot2::aes(
    size = 3.88,
    angle = 0,
    hjust = 0.5,
    vjust = 0.5,
    alpha =NA,
    family = "",
    fontface = 1,
    lineheight = 1.2,
    !!!auto_contrast
  )
)
```

and then calling for example `stat_prop(geom = GeomTextAutoContrast)`, I still get `Caused by error:
! Can't subset `.data` outside of a data mask context.`

It is like `after_scale()` is covered only when call through a global mapping, but not when used as a `default_aes`

**Reply by teunbrand** [reply]
Seems like a bug, I'll file an issue

EDIT: https://github.com/tidyverse/ggplot2/issues/6135

**Reply by larmarange** [reply]
Thanks @teunbrand 

Thanks @yjunechoe for your feedback:

> The problem with using !!! to splice here is that !!! doesn't just splice, it also injects as well (and at the ggproto's construction, .data cannot be resolved). To keep the aes(.data) expression lazy, you should use something like modifyList() instead

It still problematic.

``` r
library(ggplot2)
library(ggstats)

auto_contrast <- ggplot2::aes(colour = after_scale(hex_bw(.data$fill)))

GeomTextAutoContrast <- ggproto(
  "GeomTextAutoContrast",
  ggplot2::GeomText,
  default_aes = modifyList(
    ggplot2::aes(
      size = 3.88,
      angle = 0,
      hjust = 0.5,
      vjust = 0.5,
      alpha =NA,
      family = "",
      fontface = 1,
      lineheight = 1.2)
    ,
    auto_contrast
  )
)

layer_test <- function(mapping = NULL,
                       data = NULL,
                       position = ggplot2::position_stack(0.5),
                       ...) {
  ggplot2::layer(
    geom = GeomTextAutoContrast,
    stat = StatCount,
    data = data,
    mapping = mapping,
    position = position,
    params = list(...)
  )
}

d <- as.data.frame(Titanic)

ggplot(d) +
  aes(y = Class, fill = Survived, weight = Freq, label = after_stat(count)) +
  geom_bar() +
  layer_test()
#> Error in `layer_test()`:
#> ! Problem while setting up geom aesthetics.
#> ℹ Error occurred in the 2nd layer.
#> Caused by error:
#> ! Can't subset `.data` outside of a data mask context.
```

<sup>Created on 2024-10-11 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
@larmarange Sorry, I confused your problem with another `!!!` phenomenon (couldn't delete fast enough!). As you put it should indeed hopefully work as is :D

**Reply by yjunechoe** [reply]
This is now off topic but FYI - this is the weirdness I had in mind, where `!!!`-ing an expression containing an rlang pronoun is resolved narrowly by `!!!` injection, before `eval_tidy()` gets to see it:

```r
library(dplyr)

# External vector of indices we want to select
idx <- 2:3

# Basic equivalence
identical(
  mtcars %>% slice(1, 2, 3),
  mtcars %>% slice(!!!list(1, 2, 3))
)
#> [1] TRUE

rlang::expr(
  mtcars %>% slice(!!!list(1, 2, 3))
)
#> mtcars %>% slice(1, 2, 3)

# Basic example of expected `.env$` behavior
identical(
  mtcars %>% slice(1, 2, 3),
  mtcars %>% slice(1, .env$idx)
)
#> [1] TRUE

# Fails when `!!!` intervenes between the function invoking the data-masking context (slice) and the pronoun (.env)
mtcars %>% slice(!!!list(1, .env$idx))
#> Error: object '.env' not found
```
